File: backend/run_tests.js
#!/usr/bin/env node
/**
 * Test runner script for the Node.js backend.
 * Runs all tests and generates a coverage report.
 */

const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');
const yargs = require('yargs/yargs');
const { hideBin } = require('yargs/helpers');

// Parse command line arguments
const argv = yargs(hideBin(process.argv))
  .option('verbose', {
    alias: 'v',
    type: 'boolean',
    description: 'Run tests with verbose output'
  })
  .option('coverage', {
    alias: 'c',
    type: 'boolean',
    description: 'Generate coverage report'
  })
  .option('unit', {
    alias: 'u',
    type: 'boolean',
    description: 'Run only unit tests'
  })
  .option('integration', {
    alias: 'i',
    type: 'boolean',
    description: 'Run only integration tests'
  })
  .option('e2e', {
    alias: 'e',
    type: 'boolean',
    description: 'Run only end-to-end tests'
  })
  .option('file', {
    alias: 'f',
    type: 'string',
    description: 'Specific test file to run'
  })
  .help()
  .alias('help', 'h')
  .argv;

/**
 * Run the tests with the specified options.
 * @param {Object} options - Test options
 * @returns {Promise<number>} - Exit code
 */
async function runTests(options) {
  // Base command
  const cmd = 'npx';
  const args = ['mocha'];
  
  // Add verbosity
  if (options.verbose) {
    args.push('--verbose');
  }
  
  // Add coverage
  if (options.coverage) {
    args.unshift('nyc', '--reporter=lcov', '--reporter=text');
  }
  
  // Add test files
  if (options.file) {
    args.push(options.file);
  } else if (options.unit) {
    args.push('test/**/*.test.js', '!test/integration/**', '!test/e2e/**');
  } else if (options.integration) {
    args.push('test/integration/**/*.test.js');
  } else if (options.e2e) {
    args.push('test/e2e/**/*.test.js');
  } else {
    args.push('test/**/*.test.js');
  }
  
  // Run the tests
  console.log(`Running command: ${cmd} ${args.join(' ')}`);
  
  return new Promise((resolve) => {
    const testProcess = spawn(cmd, args, {
      stdio: 'inherit',
      shell: true
    });
    
    testProcess.on('close', (code) => {
      resolve(code);
    });
  });
}

/**
 * Main function
 */
async function main() {
  try {
    const exitCode = await runTests(argv);
    process.exit(exitCode);
  } catch (error) {
    console.error('Error running tests:', error);
    process.exit(1);
  }
}

// Run the main function
main();

File: china-growth-game/economic-model/run_tests.py
#!/usr/bin/env python
"""
Test runner script for the economic model.
Runs all tests and generates a coverage report.
"""

import os
import sys
import subprocess
import argparse

def run_tests(args):
    """Run the tests with the specified options."""
    # Base command
    cmd = ["pytest"]
    
    # Add verbosity
    if args.verbose:
        cmd.append("-v")
    
    # Add coverage
    if args.coverage:
        cmd.extend(["--cov=.", "--cov-report=term", "--cov-report=html"])
    
    # Add markers
    if args.unit:
        cmd.append("-m unit")
    elif args.integration:
        cmd.append("-m integration")
    elif args.benchmark:
        cmd.append("-m benchmark")
    elif args.api:
        cmd.append("-m api")
    
    # Add specific test files
    if args.test_file:
        cmd.append(args.test_file)
    
    # Run the tests
    print(f"Running command: {' '.join(cmd)}")
    result = subprocess.run(cmd)
    
    return result.returncode

def main():
    """Parse arguments and run tests."""
    parser = argparse.ArgumentParser(description="Run tests for the economic model")
    parser.add_argument("-v", "--verbose", action="store_true", help="Increase verbosity")
    parser.add_argument("--coverage", action="store_true", help="Generate coverage report")
    parser.add_argument("--unit", action="store_true", help="Run only unit tests")
    parser.add_argument("--integration", action="store_true", help="Run only integration tests")
    parser.add_argument("--benchmark", action="store_true", help="Run only benchmark tests")
    parser.add_argument("--api", action="store_true", help="Run only API tests")
    parser.add_argument("test_file", nargs="?", help="Specific test file to run")
    
    args = parser.parse_args()
    
    # Run the tests
    return run_tests(args)

if __name__ == "__main__":
    sys.exit(main())
